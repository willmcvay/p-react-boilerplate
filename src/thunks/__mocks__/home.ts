export const homeDataFetch = jest.fn().mockReturnValue('HOME CALLED')
