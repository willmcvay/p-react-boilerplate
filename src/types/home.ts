export interface HomeItem {
  name: string
}

export interface HomeState {
  loading: boolean
  homeData: HomeItem | null
}
