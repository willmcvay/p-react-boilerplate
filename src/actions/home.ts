import { actionCreator } from '../utils/actions'
import ActionTypes from '../constants/action-types'
import { HomeItem } from '../reducers/home'

export const homeRequestData = actionCreator<void>(ActionTypes.HOME_REQUEST_DATA)
export const homeLoading = actionCreator<boolean>(ActionTypes.HOME_LOADING)
export const homeReceiveData = actionCreator<HomeItem | undefined>(ActionTypes.HOME_RECEIVE_DATA)
export const homeClearData = actionCreator<null>(ActionTypes.HOME_CLEAR_DATA)
