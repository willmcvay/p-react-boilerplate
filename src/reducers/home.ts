import { Action } from '../types/core'
import { isType } from '../utils/actions'
import { homeLoading, homeReceiveData, homeClearData } from '../actions/home'

export interface HomeItem {
  data: {
    children: {
      data: {
        id: string
        title: string
        url: string
      }
    }[]
  }
}

export interface HomeState {
  loading: boolean
  homeData: HomeItem | null
}

export const defaultState: HomeState = {
  loading: false,
  homeData: null
}

const homeReducer = (state: HomeState = defaultState, action: Action<any>): HomeState => {
  if (isType(action, homeLoading)) {
    return {
      ...state,
      loading: action.data
    }
  }

  if (isType(action, homeReceiveData)) {
    return {
      ...state,
      loading: false,
      homeData: action.data || null
    }
  }

  if (isType(action, homeClearData)) {
    return {
      ...state,
      loading: false,
      homeData: action.data
    }
  }

  return state
}

export default homeReducer
