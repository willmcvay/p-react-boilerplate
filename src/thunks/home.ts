import { Dispatch } from 'redux'
import fetcher from '../utils/fetcher'
import { URLS } from '../constants/api'
import { homeLoading, homeReceiveData } from '../actions/home'
import { HomeItem } from '../types/home'

export const homeDataFetch = () => async (dispatch: Dispatch) => {
  try {
    dispatch(homeLoading(true))

    const redditData = await fetcher({ url: URLS.react, method: 'GET' })

    if (redditData) {
      dispatch(homeReceiveData(redditData as HomeItem))
    }
  } catch (err) {
    console.error(err.message)
  }
}
