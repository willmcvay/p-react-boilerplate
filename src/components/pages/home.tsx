import * as React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { ReduxState } from '../../types/core'
import { HomeState } from '../../reducers/home'
import { homeClearData, homeRequestData } from '../../actions/home'
import Routes from '../../constants/routes'

export interface HomeMappedActions {
  homeClearData: () => void
  homeRequestData: () => void
}

export interface HomeMappedProps {
  homeState: HomeState
}
export type HomeProps = HomeMappedActions & HomeMappedProps

export const Home: React.FunctionComponent<HomeProps> = ({ homeClearData, homeRequestData, homeState }) => (
  <div className="container">
    <h1 className="h1">React JS</h1>
    <ul className="nav">
      <li className="nav-item">
        <a className="nav-link" onClick={homeClearData}>
          Clear All
        </a>
      </li>
      <li className="nav-item">
        <a className="nav-link" onClick={homeRequestData}>
          Fetch All
        </a>
      </li>
      <li className="nav-item">
        <Link className="nav-link" to={Routes.ITEM}>
          Item Page
        </Link>
      </li>
    </ul>
    {homeState.loading ? (
      <h4 className="h4">Loading</h4>
    ) : (
      <div className="list-group">
        {homeState.homeData &&
          homeState.homeData.data.children.map(child => (
            <div key={child.data.id}>
              <a className="list-group-item list-group-item-action" target="_blank" href={child.data.url}>
                {child.data.title}
              </a>
            </div>
          ))}
      </div>
    )}
  </div>
)

const mapStateToProps = (state: ReduxState): HomeMappedProps => ({
  homeState: state.home
})

const mapDispatchToProps = (dispatch: any): HomeMappedActions => ({
  homeClearData: () => dispatch(homeClearData(null)),
  homeRequestData: () => dispatch(homeRequestData())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home)
