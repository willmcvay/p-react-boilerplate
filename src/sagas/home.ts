import fetcher from '../utils/fetcher'
import { URLS } from '../constants/api'
import { homeLoading, homeReceiveData } from '../actions/home'
import { HomeItem } from '../reducers/home'
import { put, call, fork, takeEvery, all } from '@redux-saga/core/effects'
import ActionTypes from '../constants/action-types'

export const homeDataFetch = function*() {
  yield put(homeLoading(true))

  try {
    const redditData: HomeItem | undefined = yield call(fetcher, { url: URLS.react, method: 'GET' })

    yield put(homeReceiveData(redditData))
  } catch (err) {
    // TODO - should dispatch to error handler state
    console.error(err.message)
  }
}

export const homeDataListen = function*() {
  yield takeEvery(ActionTypes.HOME_REQUEST_DATA, homeDataFetch)
}

const homeSagas = function*() {
  yield all([fork(homeDataListen)])
}

export default homeSagas
