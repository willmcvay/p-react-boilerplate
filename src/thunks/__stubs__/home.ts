import { HomeItem } from '../../reducers/home'

export const homeDataStub: HomeItem = {
  data: {
    children: [
      {
        data: {
          id: 'someid',
          title: 'Title',
          url: 'http://something.com'
        }
      }
    ]
  }
}
