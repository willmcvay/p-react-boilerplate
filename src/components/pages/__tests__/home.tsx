import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Home, HomeProps } from '../home'
import { homeDataStub } from '../../../sagas/__stubs__/home'

const props: HomeProps = {
  homeState: {
    loading: false,
    homeData: homeDataStub
  },
  homeClearData: jest.fn(),
  homeRequestData: jest.fn()
}

describe('Home', () => {
  it('should match a snapshot', () => {
    expect(toJson(shallow(<Home {...props} />))).toMatchSnapshot()
  })

  it('should dispatch homeClearData onClick', () => {
    shallow(<Home {...props} />)
      .find('a')
      .first()
      .simulate('click')

    expect(props.homeClearData).toHaveBeenCalledTimes(1)
  })

  it('should dispatch homeRequestData onClick', () => {
    shallow(<Home {...props} />)
      .find('a')
      .at(1)
      .simulate('click')

    expect(props.homeRequestData).toHaveBeenCalledTimes(1)
  })
})
