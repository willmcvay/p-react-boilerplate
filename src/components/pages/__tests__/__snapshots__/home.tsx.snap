// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match a snapshot 1`] = `
<div
  className="container"
>
  <h1
    className="h1"
  >
    React JS
  </h1>
  <ul
    className="nav"
  >
    <li
      className="nav-item"
    >
      <a
        className="nav-link"
        onClick={[MockFunction]}
      >
        Clear All
      </a>
    </li>
    <li
      className="nav-item"
    >
      <a
        className="nav-link"
        onClick={[MockFunction]}
      >
        Fetch All
      </a>
    </li>
    <li
      className="nav-item"
    >
      <Link
        className="nav-link"
        to="/item/"
      >
        Item Page
      </Link>
    </li>
  </ul>
  <div
    className="list-group"
  >
    <div
      key="someid"
    >
      <a
        className="list-group-item list-group-item-action"
        href="http://something.com"
        target="_blank"
      >
        Title
      </a>
    </div>
  </div>
</div>
`;
